### PR TITLE
[READY] Runechat: 'Runescape-like' Chat Message Overlays (It's back!) 

### PR DESCRIPTION
**Pull request by bobbah**
### Technical documentation
[HackMD Documentation](https://hackmd.io/wUGfnnyZSaiypxaNAvgWCA?view)

### Big questions you will all have

- **Will this be toggleable?**: Yes. [However, in following with a request from Oranges](https://github.com/tgstation/tgstation/pull/50608#issuecomment-618188253), this toggle functionality will be present for six months from April 23, 2020 (to be reviewed October 23, 2020) at which point the toggle will be reconsidered. If many players have disabled the feature, Runechat may be removed from the codebase, or the toggle itself may be removed if players generally favor the system.
- **How does this differ from the BeeStation PR up right now?**: This system is based on ALL messages heard, BeeStation only shows messages that are said by living mobs. You won't see messages from machines for example in BeeStation's PR.
- **How will messages be colored? Will this hurt changelings or people hiding their name?**: They will be colored based on name so that you can distinguish between sources of messages around you. Care will be taken so that changelings and folks of the like are not ousted by this system.
- **Can we have more fonts?**: Because of how BYOND renders fonts, it is difficult to have any fonts other than the one font used in the screenshots as they will generally appear blurry. I am exploring adding additional fonts to support different typefaces as I think this would be a great addition.

### The good stuff
In collaboration with Burger from BurgerStation, and with some helpful tips sprinkled on top from the folks at Goonstation, I have created from scratch a new Runescape-like chat message overlay system. It shows messages above players while they're talking in-game.

### To be done
- [x] Foundation of system for showing text on map
- [x] Add support for styling text when players yell or use megaphones, general support for existing goonchat classes that we use for language
- [x] Add colors to make players distinct 
- [x] Add preference option for turning on/off the display of these images on a per-client basis
- [x] Remove any links from messages to prevent auto anchors
- [x] Fix AI viewing messages over camera
- [x] Improve spacing between sequential messages, annoyingly maptext doesn't have margin-top or margin-bottom so this will probably take some tweaking
- [x] Determine if having a message instance per client is too expensive (ideally I wanted to have a single message per language of a given message with refs per client but with how the say pipeline is this didn't seem easily accomplished)
- [x] Determine if we should hide the whispers from talking on radio from the original user.

### Reported Bugs / Issues Needing Fixing
- [x] Radio messages are delayed on higherpop
- [x] Messages appearing under cameras and held potted plants
- [x] Holopad appending 'Says "' to text being shown
- [x] Can see chat when in deep crit / deaf
- [x] Messages rotate with the atom speaking, need to not rotate
- [x] Can't see messages when inside a machine like the nanite chamber

### WIP Media
dreamseeker_2020-04-22_23-25-53
dreamseeker_2020-04-22_23-33-16
dreamseeker_2020-04-22_23-26-20
dreamseeker_2020-04-22_23-27-07
dreamseeker_2020-04-22_23-28-26
dreamseeker_2020-04-22_23-28-46
dreamseeker_2020-04-22_23-28-58
dreamseeker_2020-04-22_23-29-19
dreamseeker_2020-04-22_23-30-06

## Why It's Good For The Game

Makes local chat a lot easier to see, helps to alleviate pressure from highpop when radio chatter fills the chat window rapidly.

## Changelog
:cl: bobbahbrown
add: Added a new chat message overlay system; now you can see all those wacky messages people are shouting at you with ease! Welcome to the Runechat era.
/:cl:

### CODE REVIEW

**Comment by MrDoomBringer** [comment]
You should make the text scale to be smaller proportional to message length to a certain extent

**Comment by MrDoomBringer** [comment]
Is it possible to make the text scale to be smaller proportional to message length? To a certain extent obviously
Also, based pr

**Comment by wesoda25** [comment]
Holy shit

**Comment by bobbahbrown** [comment]
> Is it possible to make the text scale to be smaller proportional to message length? To a certain extent obviously
> Also, based pr

everything is possible in this magic kingdom, I'll look into it!

**Comment by TheVekter** [comment]
I was going to come on and bitch about how I don't like this and I think it'll make the screen cluttered but it actually looks really nice.

**Comment by tralezab** [comment]
Remove speech bubbles and let the font change depending on what kind of speech bubble they would have 💅

**Comment by bobbahbrown** [comment]
> Remove speech bubbles and let the font change depending on what kind of speech bubble they would have 💅

Added info to the FAQ at the top of this PR related to this, but I need to find additional fonts that are open-source and friendly to being downscaled to 7px~ as BYOND is very particular with antialiasing resulting in most fonts looking very blurry at the size needed for the chat system.

I can take a peek at hiding chat bubbles for those who are using the chat system, they won't be removed outright as people can opt out of the messages being shown.

**Comment by CthulhuOnIce** [comment]
> Added info to the FAQ at the top of this PR related to this, but I need to find additional fonts that are open-source and friendly to being downscaled to 7px~ as BYOND is very particular with antialiasing resulting in most fonts looking very blurry at the size needed for the chat system.


I tried this, it's hell

**Comment by ayebee123** [comment]
trimming fire axes for free trade me in deep space

**Comment by Cyberboss** [comment]
Good PR

**Comment by necromanceranne** [comment]
buying gf

File: code/datums/chatmessage.dm
Comment (Cyberboss) [inline code/datums/chatmessage.dm:38]: Should be `CRASH` no?

**Review by Cyberboss — CHANGES_REQUESTED** [review]

**Comment by gitbirb** [comment]
if your in dark areas, does it still show the floaty speech to others? if so, that'd be pretty annoying for stealthy speaking from dark areas

File: code/datums/chatmessage.dm
Comment (MrPerson) [inline code/datums/chatmessage.dm:74]: You don't have to use a callback here. You can just do
```
animate(message, alpha = 255, time = CHAT_MESSAGE_SPAWN_TIME)
animate(alpha = 0, time = CHAT_MESSAGE_EOL_FADE)
```

File: code/datums/chatmessage.dm
Comment (MrPerson) [inline code/datums/chatmessage.dm:3]: For future reference there's a define for SECONDS so you could have just done
`#define CHAT_MESSAGE_EOL_FADE 1 SECONDS`

File: code/datums/chatmessage.dm
Comment (MrPerson) [inline code/datums/chatmessage.dm:46]: These are byte counts, not char counts. They should be length_char() and copytext_char().

File: code/datums/chatmessage.dm
Comment (MrPerson) [reply code/datums/chatmessage.dm:74]: Actually nvm you can't do that, disregard me. I mean you generally can but you can't here without mucking things up, the callback is best practice.

**Comment by HomocidlePancakes** [comment]
Make sure it doesn’t block clicks.

**Comment by Dorsisdwarf** [comment]
People can say some pretty long stuff in this game, how does that look on the popup?

**Comment by bobbahbrown** [comment]
> if your in dark areas, does it still show the floaty speech to others? if so, that'd be pretty annoying for stealthy speaking from dark areas

The messages are under the lighting layer, so if its pitch black on a tile the messages will be under that tile. Now, they will still show on top of a wall so that may prove to be problematic for some individuals. I can try to look into ensuring the player can see the speaker in terms of light-level prior to drawing the text to maybe help with this.

**Comment by bobbahbrown** [comment]
> Make sure it doesn’t block clicks.

Will double check it doesn't block clicks today.

**Comment by bobbahbrown** [comment]
> People can say some pretty long stuff in this game, how does that look on the popup?

I'm playing around with an initial value for message limiting, I'll toss some screenshots up after I have breakfast of what that looks like. It just cuts off the message and tosses on an elipses ( ... ) when it has to cut the message.

**Comment by Angustmeta** [comment]
Exciting and much needed feature imo. Tested the PR a bit. I feel the current font size is too big in all cases. It would be nice if intercoms got some text as well when standing near, even if only the first 10 letters. Goon has speech popup icon from intercoms. 

### Dark:


### Spam test:

**Comment by bobbahbrown** [comment]
@Angustmeta Thank you for the awesome feedback, I'm looking into making changes related to what you have above now. Just starting back onto this PR as I've had my breakfast and done my morning routine.

**Comment by bobbahbrown** [comment]
color moment
image

**Comment by ATH1909** [comment]
blessed bee

**Comment by BeeSting12** [comment]
Thumbs up for the toggle option, it looks like a cool feature but I feel like it would clutter the screen.

**Comment by Iamgoofball** [comment]
Don't get used to it, I'm removing the toggle once the PR is merged.

**Comment by optimumtact** [comment]
Adding a toggle for this is not a good idea. I personally will not merge it in such a state.

I recommend peoples concerns about that are mollified by better spam handling and faster aging out of messages.

and at some level, ignored because it's a disruption to the existing pattern.

**Comment by factoryman942** [comment]
can i make it wave and change colours like rs chat

**Comment by bobbahbrown** [comment]
Added exponential decay of messages based on two factors (position in message stack and cumulative height of messages), should help to reduce fatigue from people trying to fill your screen.

[Video of that here](https://cdn.greyti.de/IFWnOxbTaU206.webm)

**Comment by bobbahbrown** [comment]
In accordance with Orange's request for a non-permanent toggle functionality, and in following his suggested compromise, the toggle pref for Runechat will be present for six months from April 23, 2020 (today). This leaves it to be reviewed on October 23, 2020, at which point the toggle may be removed if the opinion of players are generally favorable or if this is not the case then the feature itself may be removed. I have added a statement reflecting this to the main body of the PR.

@Iamgoofball I would ask that you respect this; however, I am unable to stop you from attempting to change the underlying preferences.

**Comment by Marmio64** [comment]
toggle is good, no toggle is bad

**Review by spookydonut — DISMISSED** [review]

**Comment by bobbahbrown** [comment]
Most recent commit fixes the issue with radio delay, waiting on maintainer to re-apply test merge candidate

File: code/datums/chatmessage.dm
Comment (Rohesie) [inline code/datums/chatmessage.dm:14]: Put this in the client definition, please, else it breaks the LangServ go-to-definition feature, and it's harder to check for variables.
Also, think you could turn it into a lazylist?

File: code/datums/chatmessage.dm
Comment (Rohesie) [inline code/datums/chatmessage.dm:103]:
```suggestion
		for(var/msg in owned_by.seen_messages[message_loc])
			var/datum/chatmessage/m = msg
```

File: code/datums/chatmessage.dm
Comment (Rohesie) [inline code/datums/chatmessage.dm:149]:
```suggestion
	if (!client?.prefs.chat_on_map)
```

**Comment by pireamaineach** [comment]
I am of the opinion that newer messages should be added to the bottom rather than the top, and older messages should shrink as they are shifted up.

This ought to be more intuitive for reading something that spans multiple sentences.

File: code/__DEFINES/layers.dm
Comment (Rohesie) [inline code/__DEFINES/layers.dm:91]: This should be between `EMISSIVE_LAYER` and `EMISSIVE_BLOCKER_LAYER`

File: code/datums/brain_damage/imaginary_friend.dm
Comment (Rohesie) [inline code/datums/brain_damage/imaginary_friend.dm:154]:
```suggestion
	if (client?.prefs.chat_on_map)
		create_chat_message(speaker, message_language, raw_message, spans, message_mode)
```
These will get called a lot. Better check early and avoid proc overhead.

**Comment by bobbahbrown** [comment]
Resolved issue with messages rendering under plants, also prevented who can't hear from seeing the messages.

**Comment by bobbahbrown** [comment]
I've gone ahead and, using the feedback I've received from the survey so far, removed the ability to hear yourself on radio. This will prevent seeing a double message every time you speak on radio.

File: code/datums/chatmessage.dm
Comment (Rohesie) [inline code/datums/chatmessage.dm:53]: null the references to `owned_by`, `message_loc` and `message`

**Comment by BeeSting12** [comment]
I was cool with this until I heard that it wasn't going to be a preference. Straight up, I dislike the idea. It clutters up the screen in my opinion and makes it more difficult to see what's going on. Spam is made twice as annoying by this feature, which of course can be fixed by better spam handling and all that. I just don't see what was wrong with the original system, which doesn't clutter up the screen.

**Comment by optimumtact** [comment]
There will be a preference for a period of time, but after that, if the majority of people are either unable to adjust, or have adjusted, the preference will be removed, depending on what the data shows.

Every toggle, especially "legacy" toggles like this, result in a combinatorial downstream explosion in possible states for clients to be in, which means subtle bugs can be easily hidden by toggle configurations.

As such, and especially for "legacy" toggles like this, I do not wish to carry them in the codebase for ever, and instead want to timebound them to a period where people can either accept the feature or move on.

**Comment by steamp0rt** [comment]
will this oust cultists with `O'biadi nebora se'sma!`?

**Comment by tmtmtl30** [comment]
i believe that's whispered, so if whispers are being handled correctly, it shouldn't be a problem.

**Comment by bobbahbrown** [comment]
> will this oust cultists with `O'biadi nebora se'sma!`?

@steamp0rt if you would have normally heard the whisper then you'll see it.

That's all you have to worry about. There is no special case.

**Comment by willox** [comment]
Word wrapping can happen in the middle of words containing quotes like "I'm" so the I and M are on separate lines.

**Comment by bobbahbrown** [comment]
> Word wrapping can happen in the middle of words containing quotes like "I'm" so the I and M are on separate lines.

This sounds like a BYOND issue as BYOND handles wrapping but I'll look into it, thank you!

**Comment by saprasam** [comment]
this is the greatest PR

**Comment by RaveRadbury** [comment]
Hallucinations do not cause the text to pop up

**Comment by Rohesie** [comment]
Please make the preference round-worth instead of changing the savefiles while this is not ready for merge.
Let's not manipulate or change size of savefiles on TMs.

**Comment by bobbahbrown** [comment]
> Hallucinations do not cause the text to pop up

Fixed @RaveRadbury 

> Please make the preference round-worth instead of changing the savefiles while this is not ready for merge.
> Let's not manipulate or change size of savefiles on TMs.

Resolved @Rohesie

**Comment by Rohesie** [comment]
Alright, let us know when you are satisfied with the state of the PR.

**Comment by bobbahbrown** [comment]
Today's updates:
- Italicized font use was discontinued, whispering is now smaller and slightly darker to compensate.
- All non-mob speech is smaller, this seemed to be a common ask
- Fixed the reported bug with hallucinated text not being shown
- Added icons for radio messages to make it more clear as to when someone is speaking on radio

My extensive research into fonts today did not bring very beneficial results, it more or less confirmed that we should stick with the Small Fonts typeface due to BYOND's handling of fonts. I looked into several small fonts, as well as explored the concept of downsampling larger rendered font to attempt to avoid issues with anti aliasing but neither of the explored routes produced results better than what is currently there.

File: code/datums/chatmessage.dm
Comment (Rohesie) [inline code/datums/chatmessage.dm:90]:
```suggestion
	if (!ismob(target))
```

**Review by Rohesie — CHANGES_REQUESTED** [review]
Clean `seen_messages` on client `Destroy()`

File: code/modules/flufftext/Hallucination.dm
Comment (Rohesie) [inline code/modules/flufftext/Hallucination.dm:722]: This one is not really necessary, but it doesn't hurt.

**Review by Rohesie — DISMISSED** [review]
Code is fine. It just needs to be tested and, if the author is satisfied with the results, the preference made into the savefiles for merging.

**Comment by bobbahbrown** [comment]
@psykzz @Azarek the above fix prevents unconscious mobs from seeing messages

**Comment by bobbahbrown** [comment]
Added a pref for disabling messages from non-mobs, and adjusting the message length limit client-side. Re-enabled the saving of the pref for turning on/off Runechat.

I believe at this point given what I've seen in testing with the latest version that it is ready for final maintainer review and merge.

**Review by Rohesie — APPROVED** [review]

**Comment by MrDoomBringer** [comment]
what a based PR good work bobbah

**Pull request merged**
